### ISSUE TITLE
Subsequent Grpc calls with WinHttpHandler to an IIS hosted service fail

### ISSUE THREAD

**Issue by angelyordanov** [dotnet/runtime#60291]
### Description

I'm trying to call a Grpc service hosted in IIS over http2 from NetFramework app. After reading the relevant docs([supported platforms](https://docs.microsoft.com/en-us/aspnet/core/grpc/supported-platforms?view=aspnetcore-5.0#net-grpc-client-requirements)) and issues (https://github.com/dotnet/aspnetcore/issues/9020, https://github.com/dotnet/core/issues/5713) I understand that this should be possible if I'm using Win11/WinServer2022 with the pre release version of the System.Net.Http.WinHttpHandler nuget package (in my case 6.0.0-rc.1.21451.13).

Unfortunately this seems not to be the case, as only the first call to the service succeeds and any subsequent call fails with `Status(StatusCode="Cancelled", Detail="No grpc-status found on response. Using gRPC with WinHttp has Windows and package version requirements. See https://aka.ms/aspnet/grpc/netstandard for details.")`.

Things get even stranger when I tested combining calls made with the NetCore's SocketsHttpHandler. A call with the WinHttpHandler will fail no matter whether the first call to the service was made with WinHttp or Sockets, and it appears that the problem is somehow triggered by the state of the IIS application as restarting the host resets the issue(restarting the client has no effect). Furthermore even after the WinHttp call has failed one can successfully call the service with the Sockets handler from the same process.

Here is a table with the different setups
| Client        | Server        | Result |
| ------------- | ------------- | ------ |
|WinHttp | IIS | :x: Fail |
|WinHttp | Kestrel | :heavy_check_mark: Succeeds |
|Sockets | IIS | :heavy_check_mark: Succeeds |
|Sockets | Kestrel | :heavy_check_mark: Succeeds |

### Reproduction Steps

I've created a minimal project that reproduces the problem - https://github.com/angelyordanov/grpc-iis-winhttp-sample

1. Start the GrpcService with the IISExpress profile
2. Start either one of the console apps

### Expected behavior

There should be no exceptions and all grpc calls should succeed.

### Actual behavior

The second call made with the WinHttpHandler always fails.

### Regression?

No

### Known Workarounds

No

### Configuration

```
dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.401
 Commit:    4bef5f3dbf

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.401\
```

**Comment by antonfirsov** [comment]
This exception message is [coming from grpc-dotnet](https://github.com/grpc/grpc-dotnet/blob/afe2100cda2efed3cd2479c5390166cfa15ecfd3/src/Grpc.Net.Client/Internal/GrpcProtocolHelpers.cs#L374), I recommend to open an issue on that repo, unless @JamesNK can confirm that this is something we should address in WinHttpHandler.

**Comment by antonfirsov** [comment]
I'm can reproduce this with Win 11 22000.318 + VS 2019.

In the second call to the service [`GetResponseHeaderCharBufferLength`](https://github.com/dotnet/runtime/blob/a0824ba41098db9dd49fa5aa519c51788b09c5d0/src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs#L230) returns `0`, because the underlying WinHttp call returns `0` in `bufferLengthInBytes`:

https://github.com/dotnet/runtime/blob/a0824ba41098db9dd49fa5aa519c51788b09c5d0/src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs#L272-L278

This implies to me that WinHttp indeed thinks that there are no trailers in that response. Note that this happens comes consistently on every second call, I don't see how could there be a race condition.

Ideas:
- Check VS2022 IIS Express in the same environment.
- Update to a newer build of Win11 and/or ask the WinHttp team if they are aware of any recent `WINHTTP_QUERY_FLAG_TRAILERS` issue they fixed
- Packet capture + decrypt TLS (never done this before but should be possible with WireShark AFAIK ...)

**Comment by JamesNK** [comment]
Issue is in WinHttpHandler.

WINHTTP_OPTION_REQUIRE_STREAM_END needs to be specified to instruct WinHttp to read all the way to the end of the response (e.g. trailing headers frame). Without this option then WinHttp stops once it has read up to content-length.

Issue shows up when WinHttpHandler is used with IIS because IIS sometimes includes content-length header with response.

**Comment by antonfirsov** [comment]
Reopening to track 6.0.x backport.

**Comment by antonfirsov** [comment]
Backport PR is merged, this should be fixed with the upcoming (6.0.2 ?) update of the WinHttpHandler package.

**Comment by karelz** [comment]
Fixed in 7.0 (main) in PR #62870 and in 6.0.2 in PR #63346

**Comment by angelyordanov** [comment]
Now that .NET 6.0.2 is released shouldn't there be a 6.0.2 version for the System.Net.Http.WinHttpHandler nuget package with this fix included?

**Comment by JamesNK** [comment]
@karelz It doesn't appear that the NuGet - https://www.nuget.org/packages/System.Net.Http.WinHttpHandler/ - gets service releases. Is this expected? How should customers get this fix?

**Comment by karelz** [comment]
That is weird. @antonfirsov did we forget to update some package magic to get it released?

**Comment by antonfirsov** [comment]
Apologies. Looks like we didn't follow [the nuget servicing guideline](https://github.com/dotnet/runtime/blob/main/docs/project/library-servicing.md#how-to-service-a-library) so it missed the release. I'm sorry for this. Working on it, so we don't miss the train. Reopening to track that work.

**Comment by karelz** [comment]
Fixed in 7.0 (main) in PR #62870 and in ~6.0.2 in PR #63346~ 6.0.3 in PR #63346 and #65523 (which is WinHttpHandler NuGet version 6.0.1 to make versioning even more confusing)
